### PR TITLE
Allow directly setting a background to Cover

### LINKED ISSUE #16996: Cover: Allow directly setting a background
The `Cover` block has evolved quite a bit from its time as just a `Cover Image`, supporting a lot of interesting configurations (videos, overlays, etc).

It'd be interesting to let users select a solid color if they want to skip setting an image or video from the placeholder step. It could look something like this:

image

When background gradient options are finalized, we should revise the presentation a bit.

Also, once patterns (#16283) are further established, it'd be cool to surface a few predefined defaults here as well to show the variety of end results. It gets tricky with media as it cannot be easily previewed before user uploads content.

### LINKED ISSUE #17131: Cover block: Overlay setting is stuck inside the Cover setting in sidebar
**Describe the bug**
The "Overlay" settings is inside the "Cover Settings" panel. They both visually have the same hierarchy though. When closing the Cover Settings panel, the Overlay settings is no longer viewable. This is confusing.

**To reproduce**
1. Add a Cover block.
2. Add an image.
3. View the Cover Settings panel and notice the Overlay panel below it, but at the same hierarchical level.
4. Collapse the Cover Settings panel and observe the Overlay panel has disappeared.

**Expected behavior**
I expect the Overlay panel to stay visible when the Cover Settings panel is closed. They both have the same visual hierarchy.

**Screenshots**

overlay-panel

**Solution**
I think the best solution here is to rename the "Cover Settings" panel as mentioned in @kjellr's comment here: https://github.com/WordPress/gutenberg/pull/17041#issuecomment-523591254 to "Media."

### PR DESCRIPTION
**Pull request by senadir**
closes https://github.com/WordPress/gutenberg/issues/16996
closes #17131

## Description
This PR introduces color choices to cover block, essentially implementing https://github.com/WordPress/gutenberg/issues/16996
It does, however, introduces new changes to other components in order for this to happen.

- it adds `append` prop (can someone suggests a better name?) to `placeholder` & `media-placeholder` as a way to add components to the bottom of the placeholder, this could be useful to add other notices, control (like our case) or anything to extend it, it didn't make sense to filter it and rebuild the media placeholder all again for a simple thing.

- it adds `clearable` to `colorPalette` as a way to toggle the clearing button from the palette, this is useful for one time selections since there is no use for the clear button in the placeholder.

- it adds the colors to the cover block

- it plays around with showing/hiding the settings based on what is set (color or image or both)
so it disables focal point, position & overlay opacity if there is no image, the same original behavior is kept if no color is selected. 

## Introduced Gotchas
those gotchas are not problems in the implementation but things that now will cause a problem:

- now that we can have a color independent from the image, we don't have a way to remove the image after setting it, the only way is to delete the cover and creating a new one, I didn't implement this, there is only replacing the image.

- relating to the first gotcha, setting an image after setting the color is a bit implicit, but the work in https://github.com/WordPress/gutenberg/issues/11952 should help with this.
 
## How has this been tested?
- create a new cover
- set a color
- check the frontend

## Screenshots 
ezgif com-optimize (2)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by senadir** [comment]
@mapk I decide to start the implementation and wait for you to introduce the changes you want (label ect) before we merge this PR

**Comment by jasmussen** [comment]
Nice work, thank you for including a GIF. I only have a few moments to look at this so forgive me for not having tested the code yet. Based on said GIF, this looks like a nice feature, and it seems okay as an iteration to do as you suggest, and add image clearing at a later time. (Could just be a clear button next to the image section in the sidebar, something in that vein).

I would suggest that you put a max-width on the palette of colors to better match the two rows of 6 from the original mockup:

Screenshot 2019-08-15 at 08 32 40

Question: how hard would it be to include our contrast checker to compare the background color with the child items?

**Comment by senadir** [comment]
>Question: how hard would it be to include our contrast checker to compare the background color with the child items?
@jasmussen I haven't looked at it from that point yet, I assume this should be done only when there is a solid background, since we can't detect contrast on a picture.
also since cover uses innerBlocks, which allows any block to be used, it's going to be a bit tricky to check on them

**Comment by jasmussen** [comment]
Gotcha, thanks for looking.

**Comment by enriquesanchez** [comment]
@jasmussen @senadir thoughts on having the color palette on the same line as the `Upload` and `Media Library` buttons? With the possibility of having the last circle act as a button that displays the color picker?

63061519-0e531b00-bebb-11e9-9069-d7e75b2aa4db

**Comment by senadir** [comment]
How will this play with themes that have 10+ colors, see 2017 (the example in the gif) with 12 color. I’m however, all on board with the color picker option, however, I think the picker will indicate that you will select a color from the page, a colored wheel sounds better

**Comment by jasmussen** [comment]
It's worth noting that there are three buttons in the image placeholder at the moment:

Screenshot 2019-08-16 at 09 07 09

so the mockup is slightly out of date.

I do think that it would be worth it to revisit the overall design of placeholders entirely, it seems like there's an opportunity to apply a grid and better structure to the whole thing. But that of course lies a bit in the future. In the mean time, the max-width does feel like it boxes up a variable amount of swatches in a decent way.

**Comment by senadir** [comment]
@jasmussen should we update the cover to use the three buttons then?

**Comment by jasmussen** [comment]
Oh actually, I do apologize! The Cover does not have those three buttons, a little sleep deprivation from a 2yo that did not want to sleep last night confused me! Indeed there is only two buttons, my mistake, and for the time being that's out of scope for this PR to change. Though I do still think that pending a larger redesign of the placeholder, it's probably best to keep the swatches on a new line, with a max-width. Might want to center them inside if there are only a few, like in Enriques example.

**Comment by mtias** [comment]
> Might want to center them inside if there are only a few, like in Enriques example.

I think this will get tricky and create odd exceptions. We should try to show the full color palette.

**Comment by jasmussen** [comment]
> I think this will get tricky and create odd exceptions. We should try to show the full color palette on its own line.

I'm perhaps a bit technical in my point here. Let's say the default color palette comes with 12 colors. We center those inside a `margin: 0 auto;` container that has a max-width that matches two lines, so it looks like https://github.com/WordPress/gutenberg/pull/17041#issuecomment-521535307. 

But what if it's TwentyNineteen, which registers 5 colors?

Screenshot 2019-08-19 at 12 59 17

It should be possible to write some centering CSS so that when there are fewer than 6 swatches on a single line, they are centered. That's all, but can do without this too!

**Comment by senadir** [comment]
@jasmussen what do you mean with centred? as in having them in the same line as the rest of the buttons? CSS wise this is a bit tricky, to both have the wrapping & no wrapping
I guess we could leave them in their own line.

there is also the issue of custom color, should we keep the default behavior (text) or should we introduce a new shape or behavior like @enriquesanchez provided? or disable custom colors from that position (current behavior)

**Comment by jasmussen** [comment]
> what do you mean with centred? 

Sorry let me speak with a mockup :)

12 swatches:



17 swatches:

17-swatches

5 swatches:
 
5-swatches

---

Hmm. Actually, I'm glad I visualized this, it's not great. How about this?

17 swatches:



5-swatches

That is: the swatch container is inline-flex and centered and a max-width. Which means it is only as wide as the number of swatches inside. But swatches inside are left aligned. Makes sense?

**Comment by senadir** [comment]
I prefer centred but that's just a personal preference, nothing to back it up with, especially since everything else is centred, it feels weird to have left align in the center

**Comment by senadir** [comment]
@jasmussen can I get a confirm on this so I can continue? 

1. Left align or center?
2. what should we do with custom colors? a simple text or a custom button.

**Comment by mtias** [comment]
Let's do center and no custom color link for now.

**Comment by senadir** [comment]
@mtias then it's already ready for review

**Comment by youknowriad** [comment]
Capture d’écran 2019-08-21 à 4 46 07 PM

The "Cover Settings" panel is empty when I choose a color. Should we gather all the background settings under the same panel? How can I switch to an image background after choosing a color?

**Comment by youknowriad** [comment]
Mmm it looks like I can click the "edit" button in the toolbar to add an image. It doesn't feel very obvious. Should we rethink some of these controls @jasmussen @mtias ?

**Comment by senadir** [comment]
@youknowriad the image replace workflow should be addressed in another PR, @mtias already opened an issue about this and there is work that is being done on it by @draganescu.

The logic in the control get a bit tricky & complicated as you can see, I would really appreciate any feedback there in order to make it more straightforward 
https://github.com/WordPress/gutenberg/blob/39f5590cb125c998d3145e3e48696e15aecb335a/packages/block-library/src/cover/edit.js#L170-L232

**Comment by senadir** [comment]
I've added a clear button to remove the background image/video set at the moment, I'm not sure about the design so any thought are welcome @jasmussen @mapk @kjellr 
image

File: packages/block-library/src/cover/editor.scss
Comment (senadir) [inline packages/block-library/src/cover/editor.scss]: @youknowriad I feel like adding styles outside the root selector is bad behavior but I'm not sure how I can do this in a clean way other then injecting a style prop to the component

**Comment by kjellr** [comment]
> I've added a clear button to remove the background image/video set at the moment, I'm not sure about the design so any thought are welcome @jasmussen @mapk @kjellr

Thanks, @senadir. 

I think we should probably look to some other "Clear" controls for inspiration here. Here's one of the most common examples: 

Screen Shot 2019-08-21 at 2 17 31 PM

Using that as inspiration, I'd definitely place that to the end of the panel here. Not sure about alignment though: 

Frame 2

_(It's a dream of mine to align these "Reset" controls someday)_

Similarly, with this change, does it make sense to rename that panel from "Cover Settings" to "Image" (or Video), since the settings in there only concern the media?

**Comment by mapk** [comment]
Found a quick bug: The accordion headings when selecting a color background aren't jiving very well. When minimizing "Cover settings" it also minimizes "Overlay". These both seem to be on the same level of hierarchy though, so visually that shouldn't happen.

**Example:**

accordion

**Comment by mapk** [comment]
Regarding the "Clear Image" button suggestion, we also have something like this in the Document Inspector for the Featured Image. Instead of a button, it's using a red link. 

**Example:**

featured

After looking at this more, I think I prefer the default button still.

----
> Similarly, with this change, does it make sense to rename that panel from "Cover Settings" to "Image" (or Video), since the settings in there only concern the media?

I agree here. Let's do that and also remove the "Overlay" setting from within the Cover Settings panel.

**Comment by senadir** [comment]
@kjellr @mapk one thing to note about the mockup is that it only handles the case of an image as a background, having a video as a background doesn't render any settings (except colors) so I think we need some context there, see image below
Cover settings showing color setting and & empty cover settings

**Comment by senadir** [comment]
https://github.com/WordPress/gutenberg/pull/17041/commits/48dd9a349333b11c132c49355e7f47619634155a fixes #17131

**Comment by senadir** [comment]
https://github.com/WordPress/gutenberg/pull/17041/commits/6d38f41b77a7ab39d908d08d49f27e7352f7dfb7 fixes the reset button, I still need comments regarding when we set a video as a background

video:
image

image:
image

image:

**Comment by jasmussen** [comment]
Okay, I took this for a good long spin so we can ship it. Thanks again for your work.

Here's TwentyNineteen:

cover 1

Here it is, compared to the background properties of a group block:

group

Here's TwentySeventeen:

twentyseventeen

☝️ Don't mind the big font on the placeholder buttons, though, that's the fault of TwentySeventeen and/or the lack of ShadowDOM for in-canvas UI elements. Separate issue, in other words.

Works well, but let's adjust the max-width on that swatch container so it's 6 rows of swatches on two lines, rather than 7 on one and 5 on another. Here's what's in the branch, with a max-width of 300px:

Screenshot 2019-08-22 at 12 24 20

And here's the same with a max-width of 290, just arbitrarily chosen:

Screenshot 2019-08-22 at 12 24 36

Much better.

Here's adding an image, which as far as I recall is much the same as before:

images

There was a question about video, so here's a video. On Cloudup separately because file was too big for Github:

https://cloudup.com/cdvpsZlr5Ja

So, good to compare here with master:

https://cloudup.com/cwC5AsmYK8S

**Generally this branch looks decent** (aside from the max-width on the swatches). However, the "Overlay" panel appears to have gone missing entirely at least in my testing just now. It simply isn't present.

So to summarize:

- Restore "Overlay" sidebar panel both for image and video backgrounds.
- Set a smaller max-width on the swatch container, say 290, so it's 6 swatches on each line.

Optional:

As a small bonus feature, if it's easy, but okay if it's not, show the title of the video background in the Media settings panel. So instead of:

Screenshot 2019-08-22 at 12 44 11

We do this:

Screenshot 2019-08-22 at 12 44 46

**Review by jasmussen — CHANGES_REQUESTED** [review]
Delightful work, thank you!

I discovered one tiny issue with swatches, one bonus feature that you can optionally take a look at, and one regression: the overlay color panel has gone missing entirely from the cover with image or video backgrounds. See https://github.com/WordPress/gutenberg/pull/17041#issuecomment-523851529 for more.

I imagine it's a small typo somewhere, once that's spiffy, ship this!

File: packages/block-editor/src/components/media-placeholder/index.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/index.js]: This component is very tied to the Placeholder would it make sense to pass props not explicitly handler by the media placeholder to the placeholder component?

File: packages/block-library/src/cover/edit.js
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js]: Is there any case where IMAGE_BACKGROUND_TYPE === backgroundType and no URL exists? What is the reason this URL check is being added? 
We are doing the URL check in line 195 so I think we don't need to repeat it here.

File: packages/block-library/src/cover/edit.js
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js]: I guess we can remove the id check here, an image can not have an id without URL ( but can have a URL without an id).

File: packages/block-library/src/cover/edit.js
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js]: Should we also reset image/video related attributes?
Namely: hasParallax, dimRatio, backgroundType, focalPoint.

File: packages/components/src/placeholder/index.js
Comment (jorgefilipecosta) [inline packages/components/src/placeholder/index.js]: Instead of adding an append prop can't we pass what is going to appended as part of the children (last child)? If it is style purposes we can wrap the current children in a div and then pass another div with our appender class.

File: packages/block-library/src/cover/edit.js
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/edit.js]: We have a color palette component part of block editor that automatically adds the current theme colors to the pallet: https://github.com/WordPress/gutenberg//blob/359858da0675943d8a759a0a7c03e7b3846536f5/packages/block-editor/src/components/color-palette/index.js

I guess we can use it here and avoid the need to query colors at the block level.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @senadir thank you for taking this task 👍
I did a code review and things generally look good. I just left some questions/suggestions (nothing major).

File: packages/block-editor/src/components/media-placeholder/index.js
Comment (senadir) [reply packages/block-editor/src/components/media-placeholder/index.js]: yes we can do the same with `icon` & `notices` as well.

```
...props
```

File: packages/block-library/src/cover/edit.js
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: 👍🏻

File: packages/block-library/src/cover/edit.js
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: 👍🏻

File: packages/block-library/src/cover/edit.js
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: good point, they could cause some issue in conditions later

File: packages/block-library/src/cover/edit.js
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: 👍🏻

File: packages/components/src/placeholder/index.js
Comment (senadir) [reply packages/components/src/placeholder/index.js]: I'm not really sure I understand what you're saying?

**Comment by mapk** [comment]
I tested again and a few things came to mind:

1. Currently the sidebar button says "Clear Background" but because it's under the "Media Settings" accordion, I'd prefer it to say "Clear Media." I was a bit confused when it suggested clearing the background, because I had just some from the screen where I selected a solid color as the background. I get that it's actually an "overlay" of color, but there was still confusion. 

2. The edit icon (pencil) in the toolbar doesn't make sense anymore to add media to the block. I think we can iterate on this in a future PR though, especially in light of the "Replace media" component coming to fruition: https://github.com/WordPress/gutenberg/pull/16200. 

3. In relation to number 2, I feel like an option to add media should remain in an always visible "Media Settings" sidebar accordion panel. I get a bit lost when that disappears and had to figure out how media got added again. This can also be iterated on in another PR.

**Comment by senadir** [comment]
@mapk, I've updated the text. as for adding the media button to the sidebar, I'm not sure how we can do that design-wise, has there been any other blocks that follow this pattern, also I think we should leave it to another PR so we don't block this PR

**Comment by karmatosed** [comment]
I am just giving a design review and I think it's ok to merge as is, however, I have some iterations for later.

When you see this it's a kind of weird experience seeing 2 color pickers:

image

The text doesn't mention colors so how am I to know I can add a background color?

> Upload an image or video file, or pick one from your media library.

cc @mtias and @mapk for some ideas to iterate once this is in. My feelings are we should iterate the copy and not show 2 color pickers, hiding one in sidebar.

**Review by karmatosed — APPROVED** [review]

**Comment by jasmussen** [comment]
Before you merge, I noted a few days ago that the overlay colors are gone for when you use images: https://github.com/WordPress/gutenberg/pull/17041#issuecomment-523851529

Did you fix that? Haven't the time to test right now.

**Comment by senadir** [comment]
@jasmussen yes, which caused the problem @karmatosed mentioned

**Comment by senadir** [comment]
@karmatosed it should be fixed now in https://github.com/WordPress/gutenberg/pull/17041/commits/70270be07a6e3d817e837e254b431c66e752a522
image
image
image

**Comment by karmatosed** [comment]
Awesome, thank you!

**Review by jasmussen — APPROVED** [review]
Can confirm that the cover overlay works now!

Screenshot 2019-08-26 at 15 32 09

This needs a good rebase, but other than that, looks good!

File: packages/components/src/placeholder/index.js
Comment (jorgefilipecosta) [reply packages/components/src/placeholder/index.js]: Hi @senadir the Placeholder component already includes a mechanism for appending (everything added as a child is rendered inside the component) so having an append prop there seems redundant.
The problem was that MediaPlaceholder did not pass the children to the Placeholder component, but we can quickly solve that.
I created this commit to more easily explain what I'm suggesting https://github.com/WordPress/gutenberg/commit/a44eda82f2312abea613c9719acc6bb2130c579e. In this commit I add the color palette in the placeholder without using an append prop, I pass it as a child. With this change, we can then easily avoid the append addition.

**Comment by senadir** [comment]
this last commit https://github.com/WordPress/gutenberg/pull/17041/commits/fc073f1aa172b89d3de13b1ffc58e7e7cca81d32 applies suggestion by @jorgefilipecosta mentioned [here](https://github.com/WordPress/gutenberg/pull/17041#discussion_r317610786) and applied [here](https://github.com/WordPress/gutenberg/commit/a44eda82f2312abea613c9719acc6bb2130c579e)

File: packages/block-editor/src/components/media-placeholder/README.md
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/README.md]: We can now remove this prop documentation :)

Comment (senadir) [reply packages/block-editor/src/components/media-placeholder/README.md]: 😅

File: packages/block-library/src/cover/block.json
Comment (jorgefilipecosta) [inline packages/block-library/src/cover/block.json]: It seems the name attribute is added just as a way to show that there is some media when a video background is used.
If that's the case I would prefer if we remove the name for now, and as an improvement, we add a small video element to show there is a video selected. Or we show the URL currently selected. It seems adding a new attribute just for this is not worth it.
The current version would also need some additional styles:
image

So I would prefer if for now we just removed the name so we can advance with the PR and later we can use better UI for videos.

**Comment by SchneiderSam** [comment]
Just tested this new feature with the [RC-1](https://github.com/WordPress/gutenberg/releases/tag/v6.4.0-rc.1) but this is the result:

Screenshot_1

**Comment by senadir** [comment]
running from 6.4.RC tag, cannot reproduce, didn't check from a clean install for the official zip thou 
image

**Comment by jorgefilipecosta** [comment]
Hi @SchneiderSam, thank you for reporting the issue you are facing. I'm also not being able to reproduce the problem using the zip file and a brand new test site.
Could you provide additional information like the theme (or custom colors used if it is not public), a plugin that may impact this, the browser you are testing on, or any other information you may find relevant?

**Comment by senadir** [comment]
I tested the zip file on a fresh install using a custom theme that doesn't have colors support and worked fine showing the default colors.

**Comment by SchneiderSam** [comment]
Hi,

@senadir and @jorgefilipecosta all fine. This is a Theme issue with Generatepress. With the standard Theme, this PR works nice.

**Pull request merged**
